functional_programming_tests/test/with_dependency_injection.cpp

```diff
@@ -0,0 +1,528 @@
+  }
+
+ private:
+  std::shared_ptr<rclcpp::Node> node_;
+  std::shared_ptr<rclcpp::Service<example_srvs::srv::SetMap>> map_setter_;
+  std::shared_ptr<rclcpp::Service<example_srvs::srv::GetPath>> path_generator_;
+};
+
+/**
+ * @brief Convert a path from ROS message to a vector of positions
+ * @param msg The message to convert
+ * @return The path as a vector of positions
+ */
+std::vector<Position> parseGeneratedPath(
+    const std_msgs::msg::UInt8MultiArray& msg) {
+  // Check the path has an even number of elements
+  if ((msg.data.size() % 2) != 0) {
+    return {};
+  }
+  if (msg.data.size() == 0) {
+    return {};
+  }
+  std::vector<Position> path;
+
+  for (size_t idx = 0; idx < (msg.data.size() - 1); idx += 2) {
+    path.push_back({msg.data[idx], msg.data[idx + 1]});
+  }
+
+  return path;
+}
+
+/**
+ * @brief Create a sample cost map in a request message
+ * @return A shared pointer to the request message with the cost map
+ */
+std::shared_ptr<example_srvs::srv::SetMap::Request> make_costmap() {
+  auto const request = std::make_shared<example_srvs::srv::SetMap::Request>();
+
+  request->map = std_msgs::msg::UInt8MultiArray();
+
+  request->map.layout.dim.resize(3, std_msgs::msg::MultiArrayDimension());
+
+  request->map.layout.dim[0].label = "rows";
+  request->map.layout.dim[0].size = 8;
+  request->map.layout.dim[0].stride = 64;
+
+  request->map.layout.dim[1].label = "columns";
+  request->map.layout.dim[1].size = 8;
+  request->map.layout.dim[1].stride = 1;
+
+  request->map.layout.dim[2].label = "channel";
+  request->map.layout.dim[2].size = 1;
+  request->map.layout.dim[2].stride = 1;
+
+  request->map.data = {0, 0, 0, 0, 0, 0, 0, 0,  //
+                       0, 0, 0, 1, 0, 0, 0, 0,  //
+                       0, 0, 0, 1, 0, 0, 0, 0,  //
+                       0, 0, 1, 1, 1, 0, 0, 0,  //
+                       0, 0, 1, 0, 1, 1, 0, 0,  //
+                       0, 0, 1, 0, 0, 0, 0, 0,  //
+                       0, 0, 0, 0, 0, 0, 0, 0,  //
+                       0, 0, 0, 0, 0, 0, 0, 0};
+  return request;
+}
+
+// Create mock middleware handle
+struct MockMiddlewareHandle : public PathGenerator::MiddlewareHandle {
+  MOCK_METHOD(void, register_set_map_service, (SetMapServiceCallback),
+              (override));
+  MOCK_METHOD(void, register_generate_path_service,
+              (GeneratePathServiceCallback), (override));
+  MOCK_METHOD(void, log_info, (std::string const&), (override));
+  MOCK_METHOD(void, log_error, (std::string const&), (override));
+};
+
+TEST(PathGenerator, Ctor) {
+  // GIVEN a mock middleware handle
+  auto mw = std::make_unique<MockMiddlewareHandle>();
+
+  // THEN it should register the services
+  EXPECT_CALL(*mw, register_set_map_service(testing::_)).Times(1);
+  EXPECT_CALL(*mw, register_generate_path_service(testing::_)).Times(1);
+
+  // WHEN the path generator is constructed
+  auto const path_generator = PathGenerator{std::move(mw)};
+}
```

Comment:
So this is saying that to expect calls to `register_set_map_service` and `register_generate_path_service` once each when the mock middleware object is moved into PathGenerator?

Reply:
you got it